File: src/main/java/com/kevin/java/generic/TestCast.java
package com.kevin.java.generic;

/**
 * Created by: kevin
 * Date: 2022-07-05
 */
public class TestCast {
    static class Foo{
        <T> void isType(Class<?> type, T data) {
//            Class<T> type = (Class<T>) data.getClass(); //maybe passed to the method
            if (type.isInstance(data) ) {
                T t = (T) type.cast(data);
                System.out.println(t);
                // ...
            }
        }
    }
    public static void main(String[] args) {
        Foo foo = new Foo();
        foo.isType(Integer.class, 1);
    }
}

File: src/main/java/com/kevin/kotlin/generic/TestCast.kt
package com.kevin.kotlin.generic

/**
 * Created by: kevin
 * Date: 2022-07-05
 */
inline fun <reified T> List<*>.asListOfType(): List<T>? =
    if (all { it is T })
        this as List<T> else
        null

fun main() {
    val list: List<*> = mutableListOf("Hello")
    val asListOfType = list.asListOfType<Int>()
    var firstOrNull = asListOfType?.firstOrNull()
}